### PR TITLE
Add limitation check for AccessKey length

### PR DESCRIPTION
**Pull request by JialinXin**
Currently, the limitation is that access token total length should not longer than 4K. Throws exception after the check.

### CODE REVIEW

File: src/Microsoft.Azure.SignalR.Common/Utilities/AuthenticationHelper.cs [hunk 1]
```diff
@@ -14,6 +14,9 @@ namespace Microsoft.Azure.SignalR
 {
     internal static class AuthenticationHelper
     {
+        // Current limitaion is in Nigix for 4K
```
Comment (scottaddie) [inline src/Microsoft.Azure.SignalR.Common/Utilities/AuthenticationHelper.cs]:
```suggestion
         // Current limitation is in Nginx for 4K
```

Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Common/Utilities/AuthenticationHelper.cs]: remove the internal implementation  details from the comment, as we might change that and there is no need to expose that

File: src/Microsoft.Azure.SignalR.Common/Utilities/AuthenticationHelper.cs [hunk 2]
```diff
@@ -34,13 +37,20 @@ public static string GenerateAccessToken(string signingKey, string audience, IEn
...
                 audience: audience,
                 claims: claims,
                 expires: expire,
                 signingKey: signingKey,
                 requestId: requestId
             );
+
+            if (jwtToken.Length > MaxTokenLength)
+            {
+                throw new ArgumentException("AccessToken must not be longer than 4K");
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Utilities/AuthenticationHelper.cs]: For server connections, if the token too long, we should make sure exception throws.

For client’ negotiation, consider return “413 Request Entity Too Large” or “414 Request-URI Too Long”

Comment (JialinXin) [reply src/Microsoft.Azure.SignalR.Common/Utilities/AuthenticationHelper.cs]: Right. Update to return 413 for client which is more accurate. And tested locally with expected behavior.

File: src/Microsoft.Azure.SignalR.AspNet/Hubs/ServiceHubDispatcher.cs [hunk 3]
```diff
@@ -53,11 +53,19 @@ public override Task ProcessRequest(HostContext context)
...
             var claims = BuildClaims(context);
 
             // Redirect to Service
             var url = _endpoint.GetClientEndpoint();
-            var accessToken = _endpoint.GenerateClientAccessToken(claims);
+            try
+            {
+                accessToken = _endpoint.GenerateClientAccessToken(claims);
+            }
+            catch (ArgumentException)
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.AspNet/Hubs/ServiceHubDispatcher.cs]: this error is quite general, consider a specific exception type?

File: src/Microsoft.Azure.SignalR.AspNet/Hubs/ServiceHubDispatcher.cs [hunk 4]
```diff
@@ -53,11 +53,19 @@ public override Task ProcessRequest(HostContext context)
...
+            {
+                context.Response.StatusCode = 413;
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.AspNet/Hubs/ServiceHubDispatcher.cs]: also add some response message ?

File: src/Microsoft.Azure.SignalR.Common/Exception/AccessTokenTooLongException.cs [hunk 5]
```diff
@@ -0,0 +1,14 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+
+namespace Microsoft.Azure.SignalR.Common
+{
+    public class AccessTokenTooLongException : Exception
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Exception/AccessTokenTooLongException.cs]: `[Serializable]` exception, a sample https://github.com/aspnet/SignalR/blob/master/src/Microsoft.AspNetCore.SignalR.Common/HubException.cs#L13. 
As it is `public`, make the naming more specific, e.g. `AzureSignalRAccessTokenTooLongException`?

File: src/Microsoft.Azure.SignalR/NegotiateHandler.cs [hunk 6]
```diff
@@ -36,9 +37,9 @@ public NegotiationResponse Process(HttpContext context, string hubName)
             {
                 accessToken = _endpointProvider.GenerateClientAccessToken(hubName, claims);
             }
-            catch (ArgumentException)
+            catch (AccessTokenTooLongException)
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/NegotiateHandler.cs]: no need to catch as it simply throws again?

File: src/Microsoft.Azure.SignalR/ServiceRouteBuilder.cs [hunk 7]
```diff
@@ -33,6 +36,9 @@ public ServiceRouteBuilder(RouteBuilder routes)
             _routes = routes;
             _serviceProvider = _routes.ServiceProvider;
             _negotiateHandler = _serviceProvider.GetRequiredService<NegotiateHandler>();
+
+            var loggerFactory = _serviceProvider.GetService<ILoggerFactory>();
+            _logger = loggerFactory.CreateLogger<ServiceRouteBuilder>();
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/ServiceRouteBuilder.cs]: can be a null loggerFactory? `loggerFactory?.Create... ?? NullLogger...`

File: src/Microsoft.Azure.SignalR.Common/Exception/AccessTokenTooLongException.cs (hunk 5 above)
Comment (JialinXin) [reply src/Microsoft.Azure.SignalR.Common/Exception/AccessTokenTooLongException.cs]: Get it. Do we need to add summary comments as well?

---
In reply to: [243783574](https://github.com/Azure/azure-signalr/pull/303#discussion_r243783574) [](ancestors = 243783574)

File: src/Microsoft.Azure.SignalR/NegotiateHandler.cs [hunk 8]
```diff
@@ -8,6 +8,7 @@
 using Microsoft.AspNetCore.Http;
 using Microsoft.AspNetCore.Http.Connections;
 using Microsoft.AspNetCore.SignalR;
+using Microsoft.Azure.SignalR.Common;
 using Microsoft.Extensions.Options;
```
Comment (JialinXin) [inline src/Microsoft.Azure.SignalR/NegotiateHandler.cs]: Will revert the change in this file

File: test/Microsoft.Azure.SignalR.Common.Tests/AuthenticationHelperTest.cs [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        {
+            var claims = GenerateClaims(100);
+            var exception = Assert.Throws<AzureSignalRAccessTokenTooLongException>(() => AuthenticationHelper.GenerateAccessToken(SigningKey, Audience, claims, DefaultLifetime, string.Empty));
+
+            Assert.Equal("AccessToken must not be longer than 4K.", exception.Message);
+        }
+
+        private Claim[] GenerateClaims(int count)
+        {
+            var claims = new List<Claim>();
```
Comment (vicancy) [inline test/Microsoft.Azure.SignalR.Common.Tests/AuthenticationHelperTest.cs]: another way:
```
return Enumerable.Range(0, count).Select(s=>new Claim($"ClaimSubject{s}",$"ClaimValue{count}")).ToArray();
```

File: src/Microsoft.Azure.SignalR.AspNet/Hubs/ServiceHubDispatcher.cs [hunk 10]
```diff
@@ -25,13 +27,15 @@ internal class ServiceHubDispatcher : HubDispatcher
...
+        private readonly ILogger _logger;
 
         private IServiceEndpointProvider _endpoint;
 
-        public ServiceHubDispatcher(HubConfiguration configuration, string appName, ServiceOptions options) : base(configuration)
+        public ServiceHubDispatcher(HubConfiguration configuration, string appName, ServiceOptions options, ILoggerFactory loggerFactory) : base(configuration)
         {
             _appName = appName ?? throw new ArgumentException(nameof(appName));
             _claimsProvider = options?.ClaimsProvider;
+            _logger = loggerFactory.CreateLogger<ServiceHubDispatcher>();
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.AspNet/Hubs/ServiceHubDispatcher.cs]: `loggerFactory?.Create... ??`
